### PR TITLE
issue #22870 - base function to create a new fiscal year

### PR DESCRIPTION
**Pull request by gilmoskowitz**

### CODE REVIEW

**Comment by lokeshsh** [comment]
Unable to verify the fix: Please provide more information.

**Comment by sindhupriyaav** [comment]
Performed UAT: Observed no fiscal years and no accounting periods available by default in Pull Request quick start database. Please confirm the fix.

File: foundation-database/public/functions/deleteaccountingyearperiod.sql [hunk 1]
```diff
@@ -1,33 +1,33 @@
...
 
---  Check to make sure that the passed yearperiod is not closed
-  IF ( ( SELECT yearperiod_closed
-         FROM yearperiod
-         WHERE (yearperiod_id=pPeriodid) ) ) THEN
-    RETURN -1;
+  IF (SELECT yearperiod_closed
+        FROM yearperiod
+       WHERE yearperiod_id = pPeriodid) THEN
+    RAISE EXCEPTION 'Cannot delete a closed Fiscal Year. [xtuple: deleteAccountingYearPeriod, -1]';
```
Comment (bendiy) [inline foundation-database/public/functions/deleteaccountingyearperiod.sql]: Do we allow deleting the current Fiscal Year/Period? It won't be closed yet, but could have transactions in it.

File: foundation-database/public/functions/createfiscalyear.sql [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    _increment := INTERVAL '3 month';
+  ELSE
+    RAISE EXCEPTION 'Cannot create a fiscal year with style % [xtuple: createFiscalYear, -1, %, %]',
+                    pStyle, pStartDate, pStyle;
+  END IF;
+
+  WHILE _current < pStartDate + INTERVAL '1 year' LOOP
+    _next := _current + _increment;
+    _quarter := EXTRACT(QUARTER FROM ('2000-01-01'::DATE + (_current - pStartDate + 1) * INTERVAL '1 day'));
+    _ap   := createAccountingPeriod(_current, (_next - interval '1 day')::DATE, _yp, _quarter);
```
Comment (bendiy) [inline foundation-database/public/functions/createfiscalyear.sql]: This looks like it should handle leap years fine.

**Comment by bendiy** [comment]
+1
Looks like CI failed due to timeout:
https://travis-ci.org/xtuple/xtuple/builds/149020367#L2324
```
XM.CustomerGroup test
    ◦ can be loaded with a zombie session:   ✓ can be loaded with a zombie session 
    ◦ can be created:   ✓ can be created 
    ◦ can be initialized by fetching an id from the server:   ✓ can be initialized by fetching an id from the server 
    ◦ can have its values set:   ✓ can have its values set 
    ◦ can be saved to the database: About to save XM.CustomerGroup CUPGRP0.9449984848033637
{"customers":[],"name":"CUPGRP0.9449984848033637","description":null}
<<ERROR 2016-08-01T21:47:41.854Z>> Unrecoverable exception. undefined
    at invalid (/home/travis/build/xtuple/xtuple/test/lib/crud.js:250:16)
    at l (/dev/client/build/client-code:6052:37)
    at o.Events.trigger (/dev/client/build/client-code:6095:13)
```
and
https://travis-ci.org/xtuple/xtuple/builds/149020367#L2353
```
1) can be saved to the database
    ◦ can be updated:   ✓ can be updated 
    ◦ can be re-saved to the database: About to update XM.CustomerGroup Test0.738424118142575
<<ERROR 2016-08-01T21:47:51.782Z>> Unrecoverable exception. undefined
    at null._onTimeout (/home/travis/build/xtuple/xtuple/test/lib/crud.js:297:14)
    at Timer.listOnTimeout [as ontimeout] (timers.js:121:15)
  2) can be re-saved to the database
```
Points to here:
https://github.com/xtuple/xtuple/blob/4_10_x/test/lib/crud.js#L297

Test is here:
https://github.com/xtuple/xtuple/blob/4_10_x/test/specs/customer_group.js#L60-L62
Don't know why that would fail or timeout.

File: foundation-database/public/functions/deleteaccountingyearperiod.sql (hunk 1 above)
Comment (gilmoskowitz) [inline foundation-database/public/functions/deleteaccountingyearperiod.sql]: We allow it if there are no posted transactions. Those rules are enforced by `deleteAccountingPeriod()` below.

**Comment by gilmoskowitz** [comment]
@lokeshsh - Issue 22870 says that the quickstart database should not contain a pre-defined fiscal year and that the desktop client should help create one. This pull request allows the [desktop client](https://github.com/xtuple/qt-client/issues/1332) to create a fiscal year easily.

**Pull request merged**